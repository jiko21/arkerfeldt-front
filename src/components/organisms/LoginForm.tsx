/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import React, { FC, useState } from 'react';
import { Color } from '@/const/color';
import DefaultButton from '@/components/atoms/DefaultButton';
import GithubButton from '@/components/atoms/GithubButton';
import TextForm from '@/components/atoms/TextForm';
import { EmailLoginInfo } from '@/types/loginForm';
import styled from '@emotion/styled';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

const containerStyle = css`
  align-items: center;
  background: #ffffff;
  box-shadow: -4px 4px 4px rgba(0, 0, 0, 0.25), 4px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 20px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 300px;
  min-height: 320px;
  padding: 0 40px;
  span {
    color: ${Color.BLACK};
    font-size: 18px;
    font-weight: 700;
  }
  > * {
    margin: 8px 0;
  }
`;

const formStyle = css`
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 100%;
  > * {
    margin: 4px 0;
  }
`;

const Error = styled.div`
  color: red;
  text-align: left;
  height: 12px;
  font-size: 12px;
  font-weight: 700;
  width: 100%;
`;

type Props = {
  onEmailLogin: (loginInfo: EmailLoginInfo) => void;
  onGithubLogin: () => void;
  isError: boolean;
  isLoading: boolean;
};

const LoginForm: FC<Props> = ({
  onEmailLogin,
  onGithubLogin,
  isError,
  isLoading,
}) => {
  const [email, onChangeEmail] = useState('');
  const handleChangeEmail = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChangeEmail(event.target.value);
  };

  const [password, onChangePassword] = useState('');
  const handleChangePassword = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChangePassword(event.target.value);
  };

  return (
    <div css={containerStyle}>
      {isLoading ? (
        <FontAwesomeIcon
          icon={faSpinner}
          color="paleturquoise"
          size="10x"
          spin
        />
      ) : (
        <>
          <span>You should login to continue</span>
          <GithubButton onClick={onGithubLogin} testId="github-btn" />
          <span>OR</span>
          <div css={formStyle}>
            <TextForm
              testId="email"
              type={'email'}
              placeholder={'email'}
              value={email}
              onChange={handleChangeEmail}
            />
            <TextForm
              testId="password"
              type={'password'}
              placeholder={'password'}
              value={password}
              onChange={handleChangePassword}
            />
          </div>
          <Error>{isError ? 'Sorry... try again.' : ''}</Error>
          <DefaultButton
            testId="login-btn"
            onClick={() => {
              onEmailLogin({
                email,
                password,
              });
            }}
          >
            Log In
          </DefaultButton>
        </>
      )}
    </div>
  );
};

export default LoginForm;
