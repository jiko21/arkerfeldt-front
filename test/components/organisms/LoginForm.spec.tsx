import LoginForm from '@/components/organisms/LoginForm';
import { fireEvent, render, screen } from '@testing-library/react';

describe('CheckBox.tsx', () => {
  let onEmailLoginMock: jest.Mock;
  let onGithubLoginMock: jest.Mock;
  beforeEach(() => {
    onEmailLoginMock = jest.fn();
    onGithubLoginMock = jest.fn();
  });

  afterEach(() => {
    onEmailLoginMock.mockClear();
    onGithubLoginMock.mockClear();
  });

  it('should correctly render', () => {
    const props = {
      onEmailLogin: onEmailLoginMock,
      onGithubLogin: onGithubLoginMock,
      isLoading: false,
      isError: false,
    };
    const { container } = render(<LoginForm {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should correctly render (with error)', () => {
    const props = {
      onEmailLogin: onEmailLoginMock,
      onGithubLogin: onGithubLoginMock,
      isLoading: false,
      isError: true,
    };
    const { container } = render(<LoginForm {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should correctly render (with loading)', () => {
    const props = {
      onEmailLogin: onEmailLoginMock,
      onGithubLogin: onGithubLoginMock,
      isLoading: true,
      isError: false,
    };
    const { container } = render(<LoginForm {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should correctly call onEmailLoginMock when clicked email login button', () => {
    const props = {
      onEmailLogin: onEmailLoginMock,
      onGithubLogin: onGithubLoginMock,
      isLoading: false,
      isError: false,
    };
    render(<LoginForm {...props} />);
    fireEvent.input(screen.getByTestId('email'), {
      target: {
        value: 'test',
      },
    });
    fireEvent.input(screen.getByTestId('password'), {
      target: {
        value: 'password',
      },
    });

    fireEvent.click(screen.getByTestId('login-btn'));
    expect(onEmailLoginMock).toBeCalledWith({
      email: 'test',
      password: 'password',
    });
  });

  it('should correctly call onGithubLoginMock when clicked github login button', () => {
    const props = {
      onEmailLogin: onEmailLoginMock,
      onGithubLogin: onGithubLoginMock,
      isLoading: false,
      isError: false,
    };
    render(<LoginForm {...props} />);
    fireEvent.click(screen.getByTestId('github-btn'));
    expect(onGithubLoginMock).toBeCalled();
  });
});
