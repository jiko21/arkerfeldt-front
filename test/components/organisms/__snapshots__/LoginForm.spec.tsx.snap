// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox.tsx should correctly render (with error) 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #ffffff;
  box-shadow: -4px 4px 4px rgba(0, 0, 0, 0.25),4px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 300px;
  min-height: 320px;
  padding: 0 40px;
}

.emotion-0 span {
  color: #2d2d2d;
  font-size: 18px;
  font-weight: 700;
}

.emotion-0>* {
  margin: 8px 0;
}

.emotion-1 {
  background-color: #000000;
  border: none;
  border-radius: 6px;
  color: #ffffff;
  height: 30px;
  min-width: 240px;
  width: 100%;
  -webkit-transition: 0.5s;
  transition: 0.5s;
}

.emotion-1:hover {
  -webkit-transform: translateY(-4px);
  -moz-transform: translateY(-4px);
  -ms-transform: translateY(-4px);
  transform: translateY(-4px);
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  font-weight: 700;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 img {
  height: 24px;
  margin-right: 12px;
  width: 24px;
}

.emotion-3 {
  margin-right: 8px;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
}

.emotion-4>* {
  margin: 4px 0;
}

.emotion-5 {
  border: 1px solid #2d2d2d;
  border-radius: 6px;
  height: 30px;
  min-width: 240px;
  width: 100%;
}

.emotion-5::-webkit-input-placeholder {
  color: #878686;
}

.emotion-5::-moz-placeholder {
  color: #878686;
}

.emotion-5:-ms-input-placeholder {
  color: #878686;
}

.emotion-5::placeholder {
  color: #878686;
}

.emotion-7 {
  color: red;
  text-align: left;
  height: 12px;
  font-size: 12px;
  font-weight: 700;
  width: 100%;
}

.emotion-8 {
  background-color: #2d2d2d;
  border: none;
  border-radius: 6px;
  color: #ffffff;
  font-size: 18px;
  font-weight: 700;
  height: 30px;
  min-width: 240px;
  width: 100%;
  -webkit-transition: 0.5s;
  transition: 0.5s;
}

.emotion-8:hover {
  -webkit-transform: translateY(-4px);
  -moz-transform: translateY(-4px);
  -ms-transform: translateY(-4px);
  transform: translateY(-4px);
}

<div>
  <div
    class="emotion-0"
  >
    <span>
      You should login to continue
    </span>
    <button
      class="emotion-1"
      data-testid="github-btn"
    >
      <div
        class="emotion-2"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-github fa-w-16 fa-inverse emotion-3"
          color="black"
          data-icon="github"
          data-prefix="fab"
          focusable="false"
          role="img"
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
            fill="currentColor"
          />
        </svg>
        Sign in with GitHub
      </div>
    </button>
    <span>
      OR
    </span>
    <div
      class="emotion-4"
    >
      <input
        class="emotion-5"
        data-testid="email"
        placeholder="email"
        type="email"
        value=""
      />
      <input
        class="emotion-5"
        data-testid="password"
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="emotion-7"
    >
      Sorry... try again.
    </div>
    <button
      class="emotion-8"
      data-testid="login-btn"
    >
      Log In
    </button>
  </div>
</div>
`;

exports[`CheckBox.tsx should correctly render (with loading) 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #ffffff;
  box-shadow: -4px 4px 4px rgba(0, 0, 0, 0.25),4px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 300px;
  min-height: 320px;
  padding: 0 40px;
}

.emotion-0 span {
  color: #2d2d2d;
  font-size: 18px;
  font-weight: 700;
}

.emotion-0>* {
  margin: 8px 0;
}

<div>
  <div
    class="emotion-0"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-spinner fa-w-16 fa-spin fa-10x "
      color="paleturquoise"
      data-icon="spinner"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`CheckBox.tsx should correctly render 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #ffffff;
  box-shadow: -4px 4px 4px rgba(0, 0, 0, 0.25),4px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 300px;
  min-height: 320px;
  padding: 0 40px;
}

.emotion-0 span {
  color: #2d2d2d;
  font-size: 18px;
  font-weight: 700;
}

.emotion-0>* {
  margin: 8px 0;
}

.emotion-1 {
  background-color: #000000;
  border: none;
  border-radius: 6px;
  color: #ffffff;
  height: 30px;
  min-width: 240px;
  width: 100%;
  -webkit-transition: 0.5s;
  transition: 0.5s;
}

.emotion-1:hover {
  -webkit-transform: translateY(-4px);
  -moz-transform: translateY(-4px);
  -ms-transform: translateY(-4px);
  transform: translateY(-4px);
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  font-weight: 700;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 img {
  height: 24px;
  margin-right: 12px;
  width: 24px;
}

.emotion-3 {
  margin-right: 8px;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
}

.emotion-4>* {
  margin: 4px 0;
}

.emotion-5 {
  border: 1px solid #2d2d2d;
  border-radius: 6px;
  height: 30px;
  min-width: 240px;
  width: 100%;
}

.emotion-5::-webkit-input-placeholder {
  color: #878686;
}

.emotion-5::-moz-placeholder {
  color: #878686;
}

.emotion-5:-ms-input-placeholder {
  color: #878686;
}

.emotion-5::placeholder {
  color: #878686;
}

.emotion-7 {
  color: red;
  text-align: left;
  height: 12px;
  font-size: 12px;
  font-weight: 700;
  width: 100%;
}

.emotion-8 {
  background-color: #2d2d2d;
  border: none;
  border-radius: 6px;
  color: #ffffff;
  font-size: 18px;
  font-weight: 700;
  height: 30px;
  min-width: 240px;
  width: 100%;
  -webkit-transition: 0.5s;
  transition: 0.5s;
}

.emotion-8:hover {
  -webkit-transform: translateY(-4px);
  -moz-transform: translateY(-4px);
  -ms-transform: translateY(-4px);
  transform: translateY(-4px);
}

<div>
  <div
    class="emotion-0"
  >
    <span>
      You should login to continue
    </span>
    <button
      class="emotion-1"
      data-testid="github-btn"
    >
      <div
        class="emotion-2"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-github fa-w-16 fa-inverse emotion-3"
          color="black"
          data-icon="github"
          data-prefix="fab"
          focusable="false"
          role="img"
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
            fill="currentColor"
          />
        </svg>
        Sign in with GitHub
      </div>
    </button>
    <span>
      OR
    </span>
    <div
      class="emotion-4"
    >
      <input
        class="emotion-5"
        data-testid="email"
        placeholder="email"
        type="email"
        value=""
      />
      <input
        class="emotion-5"
        data-testid="password"
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="emotion-7"
    />
    <button
      class="emotion-8"
      data-testid="login-btn"
    >
      Log In
    </button>
  </div>
</div>
`;
